processing/src/main/java/io/druid/segment/DimensionQueryHelper.java

```diff
@@ -0,0 +1,359 @@
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment;
+
+import com.google.common.base.Function;
+import com.google.common.hash.Hasher;
+import io.druid.query.aggregation.Aggregator;
+import io.druid.query.QueryDimensionInfo;
+import io.druid.query.aggregation.hyperloglog.HyperLogLogCollector;
+import io.druid.query.dimension.DimensionSpec;
+import io.druid.query.filter.DruidPredicateFactory;
+import io.druid.query.filter.ValueMatcher;
+import io.druid.query.search.search.SearchHit;
+import io.druid.query.search.search.SearchQuerySpec;
+import io.druid.query.topn.TopNParams;
+import io.druid.query.topn.TopNQuery;
+import org.apache.commons.lang.mutable.MutableInt;
+
+import java.nio.ByteBuffer;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+public interface DimensionQueryHelper<EncodedType extends Comparable<EncodedType>, EncodedTypeArray, ActualType extends Comparable<ActualType>>
+{
+  /**
+   * Get a typed column value selector (DimensionSelector, LongColumnSelector, etc.) from a ColumnSelectorFactory.
+   * @param dimensionSpec The dimension of the selector
+   * @param columnSelectorFactory Column value selector provider
+   * @return Column value selector for the dimension specified by dimensionSpec.
+   */
+  public Object getColumnValueSelector(DimensionSpec dimensionSpec, ColumnSelectorFactory columnSelectorFactory);
+
+
+  /**
+   * Get the size of a row object.
+   *
+   * The type of the row object will depend on the dimension type, e.g.:
+   *
+   * String type -> IndexedInts row object
+   * Long type -> IndexedLongs row object
+   *
+   * @param rowValues The row object to return the size of
+   * @return size of the row object
+   */
+  public int getRowSize(Object rowValues);
+
+
+  /**
+   * Get the cardinality, if possible, from a dimension value selector object.
+   *
+   * The class of the row object will depend on the dimension type, e.g:
+   *
+   * String type -> DimensionSelector
+   * Long type -> LongColumnSelector
+   *
+   * @param valueSelector The dimension value selector object
+   * @return Cardinality of the dimension value selector object, -1 if cardinality is not available.
+   */
+  public int getCardinality(Object valueSelector);
```

Comment:
The long-typed dimensions won't support cardinality check initially, I think that could be handled in a later PR, it would require additional tracking of values during indexing and some changes to the long column format on disk